src/common/src/array/list_array.rs

```diff
@@ -431,118 +376,189 @@ impl ListValue {
-        iter_elems_ref!(self, it, {
-            it.map(|datum_ref| {
-                if let Some(ScalarRefImpl::List(list_ref)) = datum_ref {
-                    list_ref.flatten_len()
-                } else {
-                    1
-                }
-            })
-            .sum()
-        })
+    /// Returns the elements in the flattened list.
+    pub fn flatten(self) -> ListRef<'a> {
+        match self.array {
+            ArrayImpl::List(inner) => ListRef {
+                array: &inner.value,
+                start: inner.offsets[self.start as usize],
+                end: inner.offsets[self.end as usize],
+            }
+            .flatten(),
+            _ => self,
+        }
     }
 
     /// Iterates over the elements of the list.
-    ///
-    /// Prefer using the macro `iter_elems_ref!` if possible to avoid the cost of enum dispatching.
-    pub fn iter(self) -> impl ExactSizeIterator<Item = DatumRef<'a>> + 'a {
-        iter_elems_ref!(self, it, { Either::Left(it) }, { Either::Right(it) })
+    pub fn iter(self) -> impl DoubleEndedIterator + ExactSizeIterator<Item = DatumRef<'a>> + 'a {
+        (self.start..self.end).map(|i| self.array.value_at(i as usize))
     }
 
     /// Get the element at the given index. Returns `None` if the index is out of bounds.
     pub fn get(self, index: usize) -> Option<DatumRef<'a>> {
-        iter_elems_ref!(self, it, {
-            let mut it = it;
-            it.nth(index)
-        })
+        if index < self.len() {
+            Some(self.array.value_at(self.start as usize + index))
+        } else {
+            None
+        }
     }
 
     pub fn memcmp_serialize(
         self,
         serializer: &mut memcomparable::Serializer<impl BufMut>,
     ) -> memcomparable::Result<()> {
         let mut inner_serializer = memcomparable::Serializer::new(vec![]);
-        iter_elems_ref!(self, it, {
-            for datum_ref in it {
-                memcmp_encoding::serialize_datum_in_composite(datum_ref, &mut inner_serializer)?
-            }
-        });
+        for datum_ref in self.iter() {
+            memcmp_encoding::serialize_datum_in_composite(datum_ref, &mut inner_serializer)?
+        }
         serializer.serialize_bytes(&inner_serializer.into_inner())
     }
 
     pub fn hash_scalar_inner<H: std::hash::Hasher>(self, state: &mut H) {
-        iter_elems_ref!(self, it, {
-            for datum_ref in it {
-                hash_datum(datum_ref, state);
-            }
-        })
+        for datum_ref in self.iter() {
+            hash_datum(datum_ref, state);
+        }
     }
 
     /// estimate the serialized size with value encoding
     pub fn estimate_serialize_size_inner(self) -> usize {
-        iter_elems_ref!(self, it, {
-            it.fold(0, |acc, datum_ref| {
-                acc + estimate_serialize_datum_size(datum_ref)
-            })
-        })
+        self.iter().map(estimate_serialize_datum_size).sum()
+    }
+
+    pub fn to_owned(self) -> ListValue {
```

Comment:
Can we implementing the trait?

Reply:
Unfortunately No. `ToOwned<Owned = ListValue>` requires `ListValue: Borrow<ListRef<'_>>`, which requires the borrow function returns `&ListRef<'_>`. It is impossible to return a `&`.

Reply:
Oh, you're right. This is somehow a GAT issue and reminds me of this library -> https://docs.rs/borrowme/latest/borrowme/.